### PR TITLE
feat(inputs.prometheus): Add option to limit body length

### LINKED ISSUE #14437: Drop responses with excessively large bodies from Prometheus exporters
### Use Case

When Telegraf runs in a Kubernetes cluster, it automatically detects newly deployed pods and scrapes metrics from their exporters.
If new pods with exporters return excessively large responses, it can cause an overflow in Telegraf's buffer. This situation affects all metrics of pods scraped by the Telegraf.
This issue is more serious in multi-tenant clusters because a pod from one cluster user can affect the pods of other users.

### Expected behavior

- Drop responses with excessively large bodies from Prometheus exporters.
- The body size threshold can be adjusted by a configuration file parameter.

### Actual behavior

- Telegraf attempts to read all responses from the target exporters. However, if the metrics exceed the buffer capacity, it is unpredictable which metrics will be dropped.

### Additional info

It seems possible to implement this by checking the value of the Content-Length response header, around the location indicated by the link below.

- https://github.com/influxdata/telegraf/blob/2b731c2d577a34ecacbb921af89a33d8b84aaecd/plugins/inputs/prometheus/prometheus.go#L430-L438

If this approach seems reasonable, I will implement it.

### PR DESCRIPTION
**Pull request by hhiroshell**
## Summary

Introduce a new config option to ensure that the length of the response body is less than the specified limit.
If the body length exceeds the limit, input.prometheus plugin doesn't send the metrics to subsequent plugins. This can prevent running out of buffer space and dropping metrics when very large data is received.

This is based on @powersj 's PR. Thank you for the pioneering work.
https://github.com/influxdata/telegraf/pull/14455

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

resolves #14437

### CODE REVIEW

File: plugins/inputs/prometheus/prometheus.go [hunk 1]
```diff
@@ -441,6 +446,10 @@ func (p *Prometheus) gatherURL(u URLAndAddress, acc telegraf.Accumulator) error
 	if err != nil {
 		return fmt.Errorf("error reading body: %w", err)
 	}
+	if p.ContentLengthLimit != 0 && len(body) > p.ContentLengthLimit {
+		p.Log.Debugf("skipping %s: content length (%d) exceeded maximum body size (%d)", u.URL, resp.ContentLength, p.ContentLengthLimit)
+		return nil
+	}
```
Comment (Hipska) [inline plugins/inputs/prometheus/prometheus.go:452]: Isn't it better to check the size before reading the whole body?

You can use `resp.ContentLength` for that.

File: plugins/inputs/prometheus/prometheus_test.go [hunk 2]
```diff
@@ -342,6 +342,28 @@ func TestPrometheusGeneratesMetricsSlowEndpointHitTheTimeoutNewConfigParameter(t
...
+	defer ts.Close()
+
+	p := &Prometheus{
+		Log:                testutil.Logger{},
+		URLs:               []string{ts.URL},
+		URLTag:             "url",
+		ContentLengthLimit: 1,
+	}
+	err := p.Init()
+	require.NoError(t, err)
```
Comment (Hipska) [inline plugins/inputs/prometheus/prometheus_test.go:359]:
```suggestion
	require.NoError(t, p.Init())
```

File: plugins/inputs/prometheus/prometheus_test.go [hunk 3]
```diff
@@ -342,6 +342,28 @@ func TestPrometheusGeneratesMetricsSlowEndpointHitTheTimeoutNewConfigParameter(t
...
+
+	var acc testutil.Accumulator
+	err = acc.GatherError(p.Gather)
+	require.NoError(t, err)
```
Comment (Hipska) [inline plugins/inputs/prometheus/prometheus_test.go:363]:
```suggestion
	require.NoError(t, acc.GatherError(p.Gather))
```

File: docs/LICENSE_OF_DEPENDENCIES.md [hunk 4]
```diff
@@ -7,6 +7,7 @@ following works:
 - code.cloudfoundry.org/clock [Apache License 2.0](https://github.com/cloudfoundry/clock/blob/master/LICENSE)
 - collectd.org [ISC License](https://github.com/collectd/go-collectd/blob/master/LICENSE)
 - dario.cat/mergo [BSD 3-Clause "New" or "Revised" License](https://github.com/imdario/mergo/blob/master/LICENSE)
+- github.com/99designs/go-keychain [MIT License](https://github.com/99designs/go-keychain/blob/master/LICENSE)
```
Comment (Hipska) [inline docs/LICENSE_OF_DEPENDENCIES.md:10]: Where does this come from?

**Review by Hipska — CHANGES_REQUESTED** [review]

File: plugins/inputs/prometheus/prometheus.go (hunk 1 above)
Comment (hhiroshell) [reply plugins/inputs/prometheus/prometheus.go:452]: Thank you for the reviews!

If exporters return responses using chunked encoding. The responses don't have a "Content-Length" header.
c.f. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding#directives

In this case, we need to read the entire body to know the actual content length.

File: docs/LICENSE_OF_DEPENDENCIES.md (hunk 4 above)
Comment (hhiroshell) [reply docs/LICENSE_OF_DEPENDENCIES.md:10]: In my local environment, the result of `make check-deps` says the line needs to be added.
Does it not reproduce in your environment?

```
$ go version
go version go1.21.5 darwin/arm64

$ make check-deps
./scripts/check-deps.sh
darwin/amd64

...(snip)...

windows/arm64
--- /var/folders/vh/0l94sp8n6hx8qwhb6wsy_xrh0000gr/T/tmp.d7GAxRLO/LICENSE_OF_DEPENDENCIES.md	2024-01-10 00:45:33
+++ /var/folders/vh/0l94sp8n6hx8qwhb6wsy_xrh0000gr/T/tmp.d7GAxRLO/HEAD	2024-01-10 00:45:33
@@ -4,0 +5 @@
+github.com/99designs/go-keychain


The docs/LICENSE_OF_DEPENDENCIES.md file does not contain the expected entries.

Lines prefixed with '+' should be added to LICENSE_OF_DEPENDENCIES.md and '-'
lines should be removed.

Include a link to the appropriate licenses for any additions.
make: *** [check-deps] Error 1
```

File: plugins/inputs/prometheus/prometheus.go (hunk 1 above)
Comment (Hipska) [reply plugins/inputs/prometheus/prometheus.go:452]: Do not confuse this with the `Content-Length` HTTP header, see https://pkg.go.dev/net/http#Response, it says how many bytes may be read from `resp.Body`.

Comment (hhiroshell) [reply plugins/inputs/prometheus/prometheus.go:452]: Understood! Thank you for pointing that out. I will try using resp.ContentLength.

File: docs/LICENSE_OF_DEPENDENCIES.md (hunk 4 above)
Comment (Hipska) [reply docs/LICENSE_OF_DEPENDENCIES.md:10]: Looks weird since that is only an indirect dependency 🤔

**Comment by powersj** [comment]
@hhiroshell,

What are the next steps on this PR? I see some test failures still.

**Comment by hhiroshell** [comment]
@powersj 

In response to [this proposal](https://github.com/influxdata/telegraf/pull/14548#discussion_r1446090207), I have been revising my implementation approach. 
It took some time due to the need for research, but now that my approach has been solidified, we will update you soon.

File: plugins/inputs/prometheus/prometheus.go (hunk 1 above)
Comment (hhiroshell) [reply plugins/inputs/prometheus/prometheus.go:452]: Hi, @Hipska 
I'm sorry for not getting back to you sooner.

In my experiment, it seems that `resp.ContentLength` does not reflect the actual body length when an exporter responds with chunked encoding.

- Add debug messages to plugins/inputs/prometheus/prometheus.go as follows:

```go
 	bodyLength := len(body)
 	p.Log.Debugf("length of actual body: (%d)", bodyLength)
	p.Log.Debugf("value of \"resp.ContentLength\": (%d)", resp.ContentLength)
```

- Build and run the code with settings that scrape metrics from an exporter returning responses with chunked encoding. The output is as below:

```
2024-01-19T09:35:30Z D! [inputs.prometheus] length of actual body: (89)
2024-01-19T09:35:30Z D! [inputs.prometheus] value of "resp.ContentLength": (-1)
```

The value of `resp.ContentLength` is `-1`, which indicates the length is unknown.
Based on these results, I will proceed with implementing the actual response length as initially planned.

※ For your reference, I will note the code of the exporter server used in the experiment.

<details><summary>metrics-server.go</summary><div>

```go
package main

import (
	"fmt"
	"net/http"
)

func main() {
	http.HandleFunc("/metrics", func(w http.ResponseWriter, r *http.Request) {
		flusher, _ := w.(http.Flusher)

		w.Header().Set("Transfer-Encoding", "chunked")

		// return sample metrics devided into three chunks
		fmt.Fprintln(w, "# HELP sample_counter_metrics")
		flusher.Flush()
		fmt.Fprintln(w, "# TYPE sample_counter_metrics")
		flusher.Flush()
		fmt.Fprintln(w, "sample_counter_metrics 100.0")
		flusher.Flush()
	})

	fmt.Println("Server started at http://localhost:8080")
	http.ListenAndServe(":8080", nil)
}

```
</div></details>

Comment (Hipska) [reply plugins/inputs/prometheus/prometheus.go:452]: Sure, go ahead then. The documentation of the field is a bit misleading then..

File: docs/LICENSE_OF_DEPENDENCIES.md (hunk 4 above)
Comment (hhiroshell) [reply docs/LICENSE_OF_DEPENDENCIES.md]: @Hipska 

After trying with the latest code, the make check-deps error no longer occurred. It seems there was something wrong with my environment. 🙇

**Comment by hhiroshell** [comment]
@powersj 

Now that all implementation is complete and there are no more CI failures, could I request a review?

File: plugins/inputs/prometheus/prometheus.go [hunk 5]
```diff
@@ -207,6 +208,10 @@ func (p *Prometheus) Init() error {
 		return err
 	}
 
+	if p.ContentLengthLimit < 0 {
+		return fmt.Errorf("content length limit must zero or larger: %d", p.ContentLengthLimit)
```
Comment (Hipska) [inline plugins/inputs/prometheus/prometheus.go:212]:
```suggestion
		return fmt.Errorf("content length limit must be zero or larger: %d", p.ContentLengthLimit)
```

**Review by Hipska — CHANGES_REQUESTED** [review]
Only this small remark

File: plugins/inputs/prometheus/prometheus.go (hunk 5 above)
Comment (hhiroshell) [reply plugins/inputs/prometheus/prometheus.go:212]: Thanks!

**Review by Hipska — APPROVED** [review]

**Review by powersj — APPROVED** [review]

**Review by srebhan — CHANGES_REQUESTED** [review]
I really think this is not the right approach. Limiting the amount of content read usually only makes sense to avoid out-of-memory issues and thus DoS. However, with the approach above, we still read _all_ (`io.ReadAll`) of the body only to afterwards reject the _already read data_.

If we want this feature we should go the whole way and use [`io.LimitedReader`](https://pkg.go.dev/io#LimitedReader) to effectively limit the memory used...

**Comment by hhiroshell** [comment]
@srebhan 
Thanks for pointing out that. Indeed, if the size of a single response is large enough, it could cause Telegraf to crash when ReadAll is called.

I'll try to change to use io.LimitedReader. Thank you for telling me!

**Comment by hhiroshell** [comment]
I apologize for the mistake made while integrating changes from the master branch, which resulted in extraneous commit logs. I will create a separate PR and kindly ask for reviews there... 🙏

**Pull request closed without merging**